include/dmic.h

```diff
@@ -0,0 +1,253 @@
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author:	Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>
+ *		Sathish Kuttan <sathish.k.kuttan@.intel.com>
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef __DMIC_H__
+#define __DMIC_H__
+
+#include <device.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * DMIC driver states
+ */
+enum dmic_state {
+	DMIC_STATE_UNINIT,	/* Uninitialized */
+	DMIC_STATE_CONFIGURED,	/* Configured */
+	DMIC_STATE_READY,	/* Ready */
+	DMIC_STATE_ACTIVE,	/* Active */
+	DMIC_STATE_PAUSED,	/* Paused */
+};
+
+/**
+ * DMIC driver trigger commands
+ */
+enum dmic_trigger {
+	DMIC_TRIGGER_STOP,	/* stop stream */
+	DMIC_TRIGGER_START,	/* start stream */
+	DMIC_TRIGGER_PAUSE,	/* pause the stream */
+	DMIC_TRIGGER_RELEASE,	/* release paused stream */
+	DMIC_TRIGGER_RESET,	/* reset */
+};
+
+/**
+ * PDM Channels LEFT / RIGHT
+ */
+enum pdm_lr {
+	PDM_CHAN_LEFT,
+	PDM_CHAN_RIGHT,
+};
+
+/**
+ * PDM Input/Output signal configuration
+ */
+struct pdm_io_cfg {
+	/* parameters global to all PDM controllers */
+
+	/* minimum clock frequency supported by the mic */
+	u32_t	min_pdm_clk_freq;
+	/* maximum clock frequency supported by the mic */
+	u32_t	max_pdm_clk_freq;
+	/* minimum duty cycle in % supported by the mic */
+	u8_t	min_pdm_clk_dc;
+	/* maximum duty cycle in % supported by the mic */
+	u8_t	max_pdm_clk_dc;
+
+	/* parameters unique to each PDM controller */
+
+	/* Bit mask to optionally invert PDM clock */
+	u8_t	pdm_clk_pol;
+	/* Bit mask to optionally invert mic data */
+	u8_t	pdm_data_pol;
+	/* Collection of clock skew values for each PDM port */
+	u32_t	pdm_clk_skew;
+};
+
+/**
+ * Buffers to be provided by the application for received data
+ * NOTE: Only Ping-Pong buffers are supported at this time
+ */
+struct stream_buf {
+	void *buffer[2];	/* pointers to the ping pong buffers */
+	size_t size;		/* size of each data buffer in bytes */
+};
+
+/**
+ * Configuration of the PCM streams to be output by the PDM hardware
+ */
+struct pcm_stream_cfg {
+	/*
+	 * if either rate or width is set to 0 for a stream,
+	 * the stream would be disabled
+	 */
+
+	/* PCM sample rate of stream A */
+	u32_t			pcm_a_rate;
+	/* PCM sample rate of stream B */
+	u32_t			pcm_b_rate;
+	/* PCM sample width of stream A */
+	u8_t			pcm_a_width;
+	/* PCM sample width of stream B */
+	u8_t			pcm_b_width;
+	/* # of channels enabled (applies to both streams) */
+	u8_t			num_chan;
+	u8_t			reserved;
+	/* ping pong buffer for stream A */
+	struct stream_buf	buffer_a;
+	/* ping pong buffer for stream B */
+	struct stream_buf	buffer_b;
+	/* callback for stream A */
+	void (*stream_a_callback)(void *buffer, size_t size);
+	/* callback for stream B */
+	void (*stream_b_callback)(void *buffer, size_t size);
+};
+
+/**
+ * Mapping/ordering of the PDM channels to logical PCM output channel
+ */
+struct pdm_chan_cfg {
```

Comment:
What about controllers which have fixed channel mapping?
Also, since there are multiple controllers could we handle them independently?

Reply:
> What about controllers which have fixed channel mapping?

I could add a channel map output parameter to return the **actual** mapping that the driver is able to create on the underlying hardware.

> Also, since there are multiple controllers could we handle them independently?

Theoretically, they could be independent if the hardware supports it (the hardware I use does not). The definition is geared towards **array** applications, so they are intended to be used as a collection of mics that are synchronized. The driver or the application on top of the driver could still separate the streams to make it look like they are independent streams.